google/cloud/aiplatform/base.py

```diff
@@ -16,13 +16,196 @@
 #
 
 import abc
-from typing import Optional
+from concurrent import futures
+import functools
+import inspect
+import threading
+from typing import Any, Callable, Dict, Optional, Sequence, Union
 
 from google.auth import credentials as auth_credentials
 from google.cloud.aiplatform import utils
 from google.cloud.aiplatform import initializer
 
 
+class FutureManager(metaclass=abc.ABCMeta):
+    """Tracks concurrent futures against this object."""
+
+    def __init__(self):
+        self.__latest_future_lock = threading.Lock()
+
+        # Always points to the latest future. All submitted futures will always
+        # form a dependency on the latest future.
+        self.__latest_future = None
+
+        # Caches Exception of any executed future. Once one exception occurs
+        # all additional futures should fail and any additional invocations will block.
+        self._exception = None
+
+    def _raise_future_exception(self):
+        """Raises exception if one of the object's futures has raised."""
+        with self.__latest_future_lock:
+            if self._exception:
+                raise self._exception
+
+    def _complete_future(self, future: futures.Future):
+        """Checks for exception of future and removes the pointer if it's still latest.
+
+        Args:
+            future (futures.Future): Required. A future to complete.
+        """
+
+        with self.__latest_future_lock:
+            try:
+                future.result()  # raises
```

Comment:
Could this `result` call contain wait-until-complete in it like LRO?

Reply:
The underlying future could be executing code that calls and waits on an LRO and that happens in this implementation. Additionally we could return an LRO as the result but for right now we are only returning either a AiPlatformResourceNounWithFutureManager class or None.